lib/engine-jit/src/engine.rs

```diff
@@ -193,130 +192,117 @@ impl JITEngineInner {
         module: &ModuleInfo,
         functions: &PrimaryMap<LocalFunctionIndex, FunctionBody>,
         function_call_trampolines: &PrimaryMap<SignatureIndex, FunctionBody>,
         dynamic_function_trampolines: &PrimaryMap<FunctionIndex, FunctionBody>,
+        custom_sections: &PrimaryMap<SectionIndex, CustomSection>,
     ) -> Result<
         (
             PrimaryMap<LocalFunctionIndex, FunctionBodyPtr>,
             PrimaryMap<SignatureIndex, FunctionBodyPtr>,
             PrimaryMap<FunctionIndex, FunctionBodyPtr>,
+            PrimaryMap<SectionIndex, SectionBodyPtr>,
         ),
         CompileError,
     > {
-        // Allocate all of the compiled functions into executable memory,
-        // copying over their contents.
-        let allocated_functions = self
-            .code_memory
-            .allocate_functions(registry, &functions)
-            .map_err(|message| {
-                CompileError::Resource(format!(
-                    "failed to allocate memory for functions: {}",
-                    message
-                ))
-            })?;
+        let function_bodies = functions
+            .values()
+            .chain(function_call_trampolines.values())
+            .chain(dynamic_function_trampolines.values())
+            .collect::<Vec<_>>();
+        let (executable_sections, data_sections): (Vec<_>, _) = custom_sections
+            .values()
+            .partition(|section| section.protection == CustomSectionProtection::ReadExecute);
+        self.code_memory.push(CodeMemory::new());
 
-        let mut allocated_function_call_trampolines: PrimaryMap<SignatureIndex, FunctionBodyPtr> =
-            PrimaryMap::new();
-        // let (indices, compiled_functions): (Vec<VMSharedSignatureIndex>, PrimaryMap<FunctionIndex, FunctionBody>) = function_call_trampolines.iter().map(|(sig_index, compiled_function)| {
-        //     let func_type = module.signatures.get(sig_index).unwrap();
-        //     let index = self.signatures.register(&func_type);
-        //     (index, compiled_function)
-        // }).filter(|(index, _)| {
-        //     !self.function_call_trampolines.contains_key(index)
-        // }).unzip();
-        for (sig_index, compiled_function) in function_call_trampolines.iter() {
-            let func_type = module.signatures.get(sig_index).unwrap();
-            let index = self.signatures.register(&func_type);
-            // if self.function_call_trampolines.contains_key(&index) {
-            //     // We don't need to allocate the trampoline in case
-            //     // it's signature is already allocated.
-            //     continue;
-            // }
-            let ptr = self
-                .code_memory
-                .allocate_for_function(registry, &compiled_function)
+        let (mut allocated_functions, allocated_executable_sections, allocated_data_sections) =
+            self.code_memory
+                .last_mut()
+                .unwrap()
```

Comment:
Why is this correct?

Reply:
Why is the unwrap correct? The immediately previous line of code is:
```
self.code_memory.push(CodeMemory::new());
```